### PR TITLE
feat: support passing registry credentials to the reaper

### LINKED ISSUE #645: [Bug]: `ReaperImage` pull doesn't use `RegistryCred` for `ContainerRequest`
### Testcontainers version

0.16.0

### Using the latest Testcontainers version?

Yes

### Host OS

Macos

### Host arch

x86

### Go version

1.19

### Docker version

```shell
Client:
 Cloud integration: v1.0.29
 Version:           20.10.20
 API version:       1.41
 Go version:        go1.18.7
 Git commit:        9fdeb9c
 Built:             Tue Oct 18 18:20:35 2022
 OS/Arch:           darwin/amd64
 Context:           default
 Experimental:      true

Server: Docker Desktop 4.13.1 (90346)
 Engine:
  Version:          20.10.20
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.7
  Git commit:       03df974
  Built:            Tue Oct 18 18:18:35 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.8
  GitCommit:        9cd3357b7fd7218e4aec3eae239db1f68a5a6ec6
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### Docker info

```shell
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.9.1)
  compose: Docker Compose (Docker Inc., v2.12.1)
  dev: Docker Dev Environments (Docker Inc., v0.0.3)
  extension: Manages Docker extensions (Docker Inc., v0.2.13)
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc., 0.6.0)
  scan: Docker Scan (Docker Inc., v0.21.0)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 17
 Server Version: 20.10.20
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 9cd3357b7fd7218e4aec3eae239db1f68a5a6ec6
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.15.49-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 7.772GiB
 Name: docker-desktop
 ID: UT6I:Z3MH:BVO7:4GHT:YRHP:MGEJ:734U:DELR:HXWX:AEIS:XLYK:VATP
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5000
  127.0.0.0/8
 Live Restore Enabled: false
```

### What happened?

I am using a private repository hosted on AWS ECR. I have the reaper image available in this private repository. I use the `aws-sdk-go-v2` to obtain the `RegistryCred` which I pass to my `testcontainers.ContainerRequest{}` struct that I use to spin up a test container. 

```go
testcontainers.GenericContainerRequest{
	ContainerRequest: testcontainers.ContainerRequest{
		RegistryCred: credentials,
		Image:        <IMAGE>,
		ReaperImage: <IMAGE>,
	},
	Started: true,
}
```

If I specify the reaper image in the `ReaperImage` field, I get the following error when starting up
```
2022/11/29 11:21:41 Failed to pull image: Error response from daemon: Head "<OMITTED>ryuk:0.3.4": no basic auth credentials, will retry
```
implying that the `RegistryCred` is not being used for pulling the reaper image.

I've verified that my `RegistryCred` is okay because if I specify the same image `...ryuk:0.3.4` in the `Image` field, `testcontainers` is able to pull the image correctly using the appropriate credentials, thus suggesting that the `RegistryCred` is being used for `Image` pulls but not for `ReaperImage` pulls.

### Relevant log output

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by mdelapenya**
- feat: support passing registry credentials to the reaper
- chore: support passing registry creds for the reaper when creating networks
- chore: support configuring the reaper as functional opts
- chor: remove unused field

## What does this PR do?
This PR is addresses the need of passing a registry credentials string to the Reaper. For that we are refactoring the `NewReaper` constructor to make it more configurable and, at the same time, deprecating it to avoid incorrect usage when creating a Reaper instance.

For the registry credentials string, we have create a functional option, as we consider it is the Go idiom for providing dynamic configuration. The functional option returns a function that receive a `containerOptions`, a private struct type with the dynamic fields to configure a reaper. In this PR we are adding the aforementioned `RegistryCredentials` string and the configurable `ImageName` string as `containerOptions`. The latter is used everywhere when creating a Reaper instance and we considered it made sense to include it in this PR, not in a follow-up.

We have created a new private constructor for the reaper, `newReaper`, which accepts functional options in a variadic manner (as last parameter). Then, the old constructor has been deprecated internally to call the private one, passing the old reaperImageName argument as the corresponding functional option. We have updated the usage of the old reaper to the new, private one. 

To set the new functional reaper options, the types allowing setting a reaper (NetworkRequest and ContainerRequest) have been modified:
- the old fields in those structs have been deprecated, in favor of the new functional options
- a new field with an array of `ContainerOption` has been added to the structs, making it possible to configure the reaper in a more dynamic (and advanced) manner.
- all usages of those fields (mostly in the tests of the library) have been updated to use the new functional opts.
- we have added tests demonstrating that the processed requests include the settings for the reaper 

## Why is it important?
We want to fix the credentials bug, but at the same time, limit the configuration of the reaper in a smooth manner, so consumer do not get affect by these transition to a more consistent reaper. You can probably identify this PR with #549 #561 and #633 because of the reaper creation. This PR could establish the basis for a more consistent reaper usage.

## Related issues

- Closes #645
- Relates #633
- Relates #561
- Relates #549

## Follow-ups
We could revisit #561 once this PR is merged. Regarding #633, if this PR gets merged we'll probably need to resolve conflicts, but keep the patterns introduced here for functional options.

### CODE REVIEW

File: docker.go [hunk 1]
```diff
@@ -961,11 +961,16 @@ func (p *DockerProvider) CreateContainer(ctx context.Context, req ContainerReque
 
 	sessionID := sessionID()
 
+	reaperOpts := containerOptions{}
```
Comment (matthewmcnew) [inline docker.go:964]: Would you want some mechanism to fall back to the deprecated `req.ReaperImage`? Perhaps initalize `reaperOpts` with the deprecated req.ReaperImage?
```suggestion
	reaperOpts := containerOptions{
	   ImageName: req.ReaperImage
	}
```

File: docker.go [hunk 2]
```diff
@@ -961,11 +961,16 @@ func (p *DockerProvider) CreateContainer(ctx context.Context, req ContainerReque
...
+	for _, opt := range req.ReaperOptions {
+		opt(&reaperOpts)
+	}
+
 	var termSignal chan bool
 	// the reaper does not need to start a reaper for itself
-	isReaperContainer := strings.EqualFold(req.Image, reaperImage(req.ReaperImage))
+	isReaperContainer := strings.EqualFold(req.Image, reaperImage(reaperOpts.ImageName))
```
Comment (matthewmcnew) [inline docker.go:971]: This should satisfy what https://github.com/testcontainers/testcontainers-go/pull/720 is attempting

File: docker.go (hunk 1 above)
Comment (mdelapenya) [reply docker.go:964]: Indeed, makes total sense

File: docker.go (hunk 2 above)
Comment (mdelapenya) [reply docker.go:971]: Thanks for checking!

File: docker.go [hunk 3]
```diff
@@ -961,11 +961,18 @@ func (p *DockerProvider) CreateContainer(ctx context.Context, req ContainerReque
 
 	sessionID := sessionID()
 
+	reaperOpts := containerOptions{
+	   ImageName: req.ReaperImage
```
Comment (mdelapenya) [inline docker.go:965]:
```suggestion
	   ImageName: req.ReaperImage,
```

**Pull request merged**